Add Ansible macros for lineinfile remediations

cipherboy opened the pull request:
#### Description:

Initial work-in-progress PR for `lineinfile`-style Ansible remediations. I'd appreciate feedback regarding initial design. I'll be updating this PR with additional Ansible macros to handle configuration directories (see the example [`in my testbed`](https://github.com/cipherboy/testbed/blob/master/ansible-lineinfile/roles/check-config-dir-duplicates/tasks/main.yml)). 

This PR adds `lineinfile` style ansible remediations as Jinja macros. In addition to the "simple" `lineinfile` remediation (a direct passthrough to turn a macro call into Ansible script), there are `one_line` and `config` based remediations. The former ensures there's only one of the specified line in the file, the latter handles building a regex from the parameter (with customizable prefix and separators regex parts). If absent, it will add the line `parameter + separator + value` to the file. 

One example is the rule `disable_host_auth` to set `HostbasedAuthentication no`  in `/etc/ssh/sshd_config`.

#### Goals:

- Add Ansible macros in this PR
- Add Bash macros to accomplish the same in a later PR.
- Add OVAL checks to check for these in a later PR.
- Extend the templating system to use these macros.

#### Rationale:

Hopefully this PR makes sense. By extending the use of macros everywhere (in #4415), we can hopefully make templating easier. This would also set us up to simplify templating later if we wanted to.

I think there's three common forms:
 - A basic config file,
 - A config file + a config directory,
 - A structured config file (ini, xml, etc).

Of these, this PR will hope to address 1 and 2 most thoroughly, though I'd like to add support for `ini` files in Ansible as well. Supporting `ini` and `xml` will be much harder in `bash` remediations. I haven't yet looked at OVAL support.

cipherboy commented on shared/macros-ansible.jinja:
```diff
@@ -0,0 +1,54 @@
+  the new line is appended to the file.
+#}}
+{{% macro ansible_one_line_in_file(msg, path, line_regex, new_line, create='no') -%}}
+{{{ ansible_simple_line_in_file("[DEDUPLICATE] " + msg, path, regex=line_regex, new_line=None, create='no', state='absent') }}}
```
Technically we could add a third one here which only checks for the if the line already exists in the file. If it does, skip these two steps. The problem is that the current Ansible roles append:

```yaml
when: ansible_virtualization_role != "guest" or ansible_virtualization_type != "docker"
```

(and YAML/Ansible doesn't support joining multiple `when` clauses with an `AND`...) so we can't easily tie the three together like that.

yuumasato replied:
I'm not sure I follow here, when the conditions are in a list won't Ansible `AND`' them?

cipherboy replied:
Ahhh I understand now. I wasn't aware that we'd automatically convert this to a list. My bad -- still getting familiar with the build system. :)

yuumasato replied:
No need to apologize. :)

jan-cerny commented on shared/macros-ansible.jinja:
```diff
@@ -0,0 +1,54 @@
+  Note that all string-like parameters are single quoted in the YAML.
+#}}
+{{% macro ansible_simple_line_in_file(msg='', path='', regex='', new_line='', create='no', state='present', with_items='', register='') -%}}
+- name: "{{{ msg }}}"
```
A small idea for improvement: In some of the rules, we use `- name: "{{{ rule_title }}}"` which means the Ansible task has the same name as the title of the rule it belongs to. We can make the `msg` parameter optional and if `msg` isn't present we can put the rule title to `name` field.

cipherboy replied:
@jan-cerny Is there a oneline for this in Jinja or do I need an `if`/`else` statement?

jan-cerny replied:
@cipherboy There is a built-in `default` filter (http://jinja.pocoo.org/docs/2.10/templates/#list-of-builtin-filters), but the if/else statement is well readable.

cipherboy replied:
I had gone with `msg or rule_title` which appears to work as well.

jan-cerny replied:
@cipherboy wow, it does!

jan-cerny commented on shared/macros-ansible.jinja:
```diff
@@ -0,0 +1,54 @@
+#}}
+{{% macro ansible_one_line_in_file(msg, path, line_regex, new_line, create='no') -%}}
+{{{ ansible_simple_line_in_file("[DEDUPLICATE] " + msg, path, regex=line_regex, new_line=None, create='no', state='absent') }}}
+{{{ ansible_simple_line_in_file(msg, path, regex=None, new_line=new_line, create=create, state='present') }}}
```
Another idea: This creates 2 independent Ansible tasks. Our build system adds tags to each task, so the output becomes little bloated (`build/rhel8/fixes/ansible/disable_host_auth.yml`) Do you think it would be feasible to use blocks (https://docs.ansible.com/ansible/latest/user_guide/playbooks_blocks.html) to merge the 2 tasks together?  I don't know if it's possible due to the indentation, but it might be worth investigating.

cipherboy replied:
Ah that'd be an interesting idea, I'll go and try it. :) Thanks @jan-cerny!

cipherboy replied:
@jan-cerny Jinja appears to provide a "indent" helper function so this is doable. :)

jan-cerny commented on shared/macros-ansible.jinja:
```diff
@@ -0,0 +1,54 @@
+{{%- endmacro %}}
+
+{{#
+  Ensure the configuration is set in a file. Note this handles generic
```
Another idea: It would be awesome if the macro is mentioned also in Developer's guide.

cipherboy replied:
Yep, I'm thinking of documenting "how to add new macros" as well.

cipherboy commented:
Sample generated Ansible content:

```yaml
<snip>

- name: Set SSH Client Alive Max Count
  block:

    - name: Deduplicate values from /etc/ssh/sshd_config
      lineinfile:
        path: /etc/ssh/sshd_config
        create: false
        regexp: (?i)^\s*ClientAliveCountMax\s+
        state: absent

    - name: Insert correct line to /etc/ssh/sshd_config
      lineinfile:
        path: /etc/ssh/sshd_config
        create: true
        line: ClientAliveCountMax {{ var_sshd_set_keepalive }}
        state: present
        validate: /usr/sbin/sshd -t -f %s
  when: ansible_virtualization_role != "guest" or ansible_virtualization_type != "docker"
  tags:

<snip>
```

cipherboy commented on linux_os/guide/services/ssh/ssh_server/sshd_print_last_log/ansible/shared.yml:
```diff
@@ -1,9 +1,6 @@
+# reboot = false
+# strategy = restrict
+# complexity = low
+# disruption = low
```
These were missing from this file so I added them.

cipherboy commented on shared/macros-ansible.jinja:
```diff
@@ -0,0 +1,128 @@
+{{%- endmacro %}}
+
+{{%- macro ansible_sshd_set(msg, parameter, value) %}}
+{{{ ansible_set_config_file(msg, "/etc/ssh/sshd_config", parameter, value=value, create="yes", prefix_regex='(?i)^\s*', validate="/usr/sbin/sshd -t -f %s") }}}
```
`sshd` configuration (unlike `ssh` client configuration) doesn't have a directory. Parameters are case sensitive, but values aren't. We also have a utility to validate the new configuration.

cipherboy commented:
Sample generated `/etc/profile` content to test directories:

```yaml
- name: Set Interactive Session Timeout
  block:

    - name: Deduplicate values from /etc/profile
      lineinfile:
        path: /etc/profile
        create: false
        regexp: ^\s*TMOUT=
        state: absent

    - name: Check if /etc/profile.d exists
      stat:
        path: /etc/profile.d
      register: _etc_profile_d_exists

    - name: Check if the parameter TMOUT is present in /etc/profile.d
      find:
        paths: /etc/profile.d
        recurse: 'yes'
        follow: 'no'
        contains: ^\s*TMOUT=
      register: _etc_profile_d_has_parameter
      when: _etc_profile_d_exists.stat.isdir is defined and _etc_profile_d_exists.stat.isdir

    - name: Remove parameter from files in /etc/profile.d
      lineinfile:
        path: '{{ item.path }}'
        create: false
        regexp: ^\s*TMOUT=
        state: absent
      with_items: '{{ _etc_profile_d_has_parameter.files }}'
      when: _etc_profile_d_has_parameter.matched

    - name: Insert correct line to /etc/profile
      lineinfile:
        path: /etc/profile
        create: true
        line: TMOUT={{ var_accounts_tmout }}
        state: present
        validate: bash -n %s
  tags:
    - accounts_tmout
```

jan-cerny commented on linux_os/guide/services/ssh/ssh_server/disable_host_auth/ansible/shared.yml:
```diff
@@ -3,9 +3,4 @@
-    dest: /etc/ssh/sshd_config
-    regexp: ^HostbasedAuthentication
-    line: HostbasedAuthentication no
+{{{ ansible_sshd_set('', "HostbasedAuthentication", "no") }}}
```
It would be better to make  `msg` an optional argument because you always pass an empty string there anyway.

cipherboy replied:
Yeah that's a good change. That and documentation are the two things I want to do still in this patch. 

Other than the two sets of rules I've changed, I don't think I'll track down everywhere these new macros can be used.

cipherboy commented:
@matejak, @jan-cerny, and @yuumasato: I think this is a good scope for this PR and would appreciate review. 

We've added some remediations using the new macros so we test that they work. We've added the macros themselves and documented them.

I'll move onto Bash macros next.

jan-cerny commented:
@cipherboy I like it very much, it's great that it's very easy to use. I have checked the generated playbooks and they look fine.

cipherboy commented on linux_os/guide/services/ssh/ssh_server/sshd_disable_user_known_hosts/ansible/shared.yml:
```diff
@@ -3,13 +3,4 @@
-    dest: /etc/ssh/sshd_config
-    regexp: ^IgnoreUserKnownHosts
-    line: IgnoreUserKnownHosts yes
-    insertbefore: ^Match
```
This is the only thing I didn't replicate. However, I can't seem to figure out *why* this was added. This configuration value isn't affected by `Match` groups...

cipherboy replied:
Note that the bash remediation doesn't care [about this](https://github.com/ComplianceAsCode/content/blob/master/linux_os/guide/services/ssh/ssh_server/sshd_disable_user_known_hosts/bash/shared.sh).

cipherboy replied:
It appears this is the relevant bug report: https://github.com/ComplianceAsCode/content/issues/3645 

I'm guessing most of our ssh configuration parameters have this or similar problems. Should we just add this to the top of the configuration file?

cipherboy replied:
Ok, this is fixed; all SSH-based thingies are added before `Match` sections.

cipherboy commented:
Most correctly, the behavior I think we want for SSH config is:

 - Remove all conflicting options
 - If there's a match group, add just before the first match group. 
 - Else add to the end of the file.

But that could be extended in a later PR.

yuumasato commented on shared/macros-ansible.jinja:
```diff
@@ -0,0 +1,152 @@
+    {{{ ansible_lineinfile("Insert correct line to " + path, path, regex=None, new_line=new_line, create=create, state='present', validate=validate, insert_after=insert_after, insert_before=insert_before)|indent }}}
+{{%- else %}}
+{{{ ansible_lineinfile("Deduplicate values from " + path, path, regex=line_regex, new_line=None, create='no', state='absent') }}}
+{{{ ansible_lineinfile("Insert correct line into " + path, path, regex=None, new_line=new_line, create=create, state='present', validate=validate) }}}
```
Is there any reason for this second call of `ansible_lineinfile` not setting `insert_after`?

cipherboy commented:
@yuumasato Ah sorry. Fixed by update. :)

jan-cerny commented:
@cipherboy Looks great!
I have tried to run a couple of test scenarios with `--remediate-using ansible` and they pass.

```
[jcerny@thinkpad tests{cipherboy-jinja-ansible-lineinfile}]$ ./test_suite.py rule --libvirt qemu:///system ssgts_rhel8 --datastream ../build/ssg-rhel8-ds.xml --remediate-using ansible rule_sshd_disable_gssapi_a
uth
Setting console output to log level INFO
INFO - The base image option has not been specified, choosing libvirt-based test environment.
INFO - Logging into /home/jcerny/scap-security-guide/tests/logs/rule-custom-2019-06-25-1247/test_suite.log
INFO - xccdf_org.ssgproject.content_rule_sshd_disable_gssapi_auth
INFO - Script comment.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script wrong_value.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script correct_value.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script line_not_there.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
[jcerny@thinkpad tests{cipherboy-jinja-ansible-lineinfile}]$ ./test_suite.py rule --libvirt qemu:///system ssgts_rhel8 --datastream ../build/ssg-rhel8-ds.xml --remediate-using ansible accounts_tmout
Setting console output to log level INFO
INFO - The base image option has not been specified, choosing libvirt-based test environment.
INFO - Logging into /home/jcerny/scap-security-guide/tests/logs/rule-custom-2019-06-25-1251/test_suite.log
INFO - xccdf_org.ssgproject.content_rule_accounts_tmout
INFO - Script comment.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script wrong_value.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script correct_value.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script supercompliance.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script line_not_there.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
[jcerny@thinkpad tests{cipherboy-jinja-ansible-lineinfile}]$ ./test_suite.py rule --libvirt qemu:///system ssgts_rhel8 --datastream ../build/ssg-rhel8-ds.xml --remediate-using ansible rule_sshd_disable_kerb_auth
Setting console output to log level INFO
INFO - The base image option has not been specified, choosing libvirt-based test environment.
INFO - Logging into /home/jcerny/scap-security-guide/tests/logs/rule-custom-2019-06-25-1257/test_suite.log
INFO - xccdf_org.ssgproject.content_rule_sshd_disable_kerb_auth
INFO - Script wrong_value.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script correct_value.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script comment.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script line_not_there.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
```

@yuumasato Do you agree it can be merged?

yuumasato commented:
> @yuumasato Do you agree it can be merged?

Yes, this looks great. 
@cipherboy Thank you.

cipherboy commented:
Cool, thanks folks!
